Add lets_split keymap

tylerwince opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->
Add a keymap for lets_split boards called tylerwince

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Keymap/layout/userspace (addition or update)

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/config.h:
```diff
@@ -0,0 +1,45 @@
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONFIG_USER_H
```
If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`?

It's simpler, and less prone to user error, so if you could make this switch?

```suggestion
#pragma once
```

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/config.h:
```diff
@@ -0,0 +1,45 @@
+*/
+
+#ifndef CONFIG_USER_H
+#define CONFIG_USER_H
```
```suggestion
```

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/config.h:
```diff
@@ -0,0 +1,45 @@
+#define RGBLIGHT_VAL_STEP 8
+
+
+#endif
```
```suggestion
```

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/config.h:
```diff
@@ -0,0 +1,45 @@
+#ifndef CONFIG_USER_H
+#define CONFIG_USER_H
+
+#include "../../config.h"
```
QMK's build script automatically includes all the `config.h` files in the build tree, so this line is redundant.

```suggestion
```

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,249 @@
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+// Layer names don't all need to be of the same length, obviously, and you can also skip them
+// entirely and just use numbers.
+#define _QWERTY 0
```
Lines 9-14 should be changed to:

```c
enum layer_names {
    _QWERTY,
    _COLEMAK,
    _DVORAK,
    _LOWER,
    _RAISE,
    _ADJUST,
};
```

This will change your layer indexes, as having them non-consecutive will increase your firmware size for no benefit or change in functionality.

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,249 @@
+  SH_GOUP,
+};
+
+// Fillers to make layering more clear
```
QMK defines these keycode aliases automatically. Lines 26-28 can be deleted.

https://github.com/qmk/qmk_firmware/blob/a120a000d481ed14a87988383be46c5a76604bb6/quantum/quantum_keycodes.h#L29-L31

```suggestion
```

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,249 @@
+};
+
+// Fillers to make layering more clear
+#define _______ KC_TRNS
```
```suggestion
```

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,249 @@
+
+// Fillers to make layering more clear
+#define _______ KC_TRNS
+#define XXXXXXX KC_NO
```
```suggestion
```

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/rules.mk:
```diff
@@ -0,0 +1,6 @@
+MOUSEKEY_ENABLE=yes
+RGBLIGHT_ENABLE=yes
+TAP_DANCE_ENABLE=yes
+ifndef QUANTUM_DIR
```
Lines 4-6 are leftover from a previous build system and no longer have any functionality. They should be deleted.

tylerwince replied:
@noroadsleft - Thanks so much for the review. I am a bit confused on this one. I use `RGBLIGHT` and `TAP_DANCE` in my keymap but are you saying that including them here doesn't matter and I should only have this file include the Makefile from the top root? (Effectively making this file only have:
```make
include ../../../../Makefile
```

tylerwince replied:
Dur -- read this wrong. 4-6. I should only include the features I use. Got it. :+1:

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,233 @@
+  _RAISE,
+  _ADJUST,
+};
+
```
Your `custom_keycodes` code block was actually fine; it was only the layer names that needed adjusting.

Everything after your Tap Dance declarations should be restored to how it was at https://github.com/qmk/qmk_firmware/blob/d8ba849efb4d24e3f80c2eb548c7a0086455589f/keyboards/lets_split/keymaps/tylerwince/keymap.c, from here to the end of the file should match Lines 29-249 of the linked revision of `keymap.c`.

tylerwince replied:
Ahh - understood. I’ll make the change when I am stationary and push.

fauxpark commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,233 @@
+ * `------------------------------------------   ------------------------------------------'
+ */
+[_QWERTY] = LAYOUT_ortho_4x12(
+   KC_TAB,        KC_Q,          KC_W,    KC_E,    KC_R,  KC_T,   KC_Y,   KC_U,  KC_I,    KC_O,    KC_P,              KC_BSPC, \
```
The backslashes at the end of these lines are not necessary. You can keep them if you want as it isn't hurting anything, but IMO it looks neater and less "hack-y".

drashna commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,233 @@
+
+};
+
+void matrix_init_user(void) {
```
Actually, you should use `layer_state_set_user` for all of the rgb stuff. It's ... better and works with any layer code. 

Also, this should be:
```c
void keyboard_post_init_user(void) {
    rgblight_enable_noeeprom();
    layer_state_set_user(layer_state);
}
```

drashna commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,233 @@
+  rgblight_sethsv(255, 0, 160);
+  #endif
+}
+void persistent_default_layer_set(uint16_t default_layer) {
```
Could you remove this function?  

It's been deprecated in favor of the `set_single_persistent_default_layer` function (which is a core function).

https://docs.qmk.fm/#/ref_functions?id=setting-the-persistent-default-layer

drashna commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,233 @@
+}
+
+void set_qwerty(void){
+  rgblight_sethsv(255, 0, 160);
```
```suggestion
  rgblight_sethsv_noeeprom(255, 0, 160);
```

fauxpark commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,239 @@
+ * |Things|1PASS | alt  | CMD  |Lower |Space |   |Space |Raise | Next | Vol- | Vol+ | Play |
+ * `------------------------------------------   ------------------------------------------'
+ */
+[_COLEMAK] = LAYOUT_ortho_4x12( \
```
```suggestion
[_COLEMAK] = LAYOUT_ortho_4x12(
```

fauxpark commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,239 @@
+ * |Things|1PASS | alt  | CMD  |Lower |Space |   |Space |Raise | Next | Vol- | Vol+ | Play |
+ * `------------------------------------------   ------------------------------------------'
+ */
+[_DVORAK] = LAYOUT_ortho_4x12( \
```
```suggestion
[_DVORAK] = LAYOUT_ortho_4x12(
```

fauxpark commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,239 @@
+ * |Things|1PASS | alt  | CMD  |Lower |Space |   |Space |Raise | Next | Vol- | Vol+ | Play |
+ * `------------------------------------------   ------------------------------------------'
+ */
+[_LOWER] = LAYOUT_ortho_4x12( \
```
```suggestion
[_LOWER] = LAYOUT_ortho_4x12(
```

fauxpark commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,239 @@
+ * |Things|1PASS | alt  | CMD  |Lower |Space |   |Space |Raise | Next | Vol- | Vol+ | Play |
+ * `------------------------------------------   ------------------------------------------'
+ */
+[_RAISE] = LAYOUT_ortho_4x12( \
```
```suggestion
[_RAISE] = LAYOUT_ortho_4x12(
```

fauxpark commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,239 @@
+ * |      |      |      |      |      |      |   |      |      |RGBwav|RGBfla|RGBtog|      |
+ * `------------------------------------------   ------------------------------------------'
+ */
+[_ADJUST] =  LAYOUT_ortho_4x12( \
```
```suggestion
[_ADJUST] =  LAYOUT_ortho_4x12(
```

tylerwince commented:
Thanks to @fauxpark, @drashna, and @noroadsleft for the close eyes on this config, it has helped me make a better configuration as well! I think I have addressed all your concerns, please take one last look and I think this is good to go.

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,239 @@
+        #ifdef RGBLIGHT_ENABLE
+          set_qwerty();
+        #endif
+        set_single_persistent_default_layer(1UL<<_QWERTY);
```
`set_single_persistent_default_layer()` just takes the layer index/name directly; no bit shifting needed.

```suggestion
        set_single_persistent_default_layer(_QWERTY);
```

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,239 @@
+        #ifdef AUDIO_ENABLE
+          PLAY_SONG(tone_colemak);
+        #endif
+        set_single_persistent_default_layer(1UL<<_COLEMAK);
```
```suggestion
        set_single_persistent_default_layer(_COLEMAK);
```

noroadsleft commented on keyboards/lets_split/keymaps/tylerwince/keymap.c:
```diff
@@ -0,0 +1,239 @@
+        #ifdef AUDIO_ENABLE
+          PLAY_SONG(tone_dvorak);
+        #endif
+        set_single_persistent_default_layer(1UL<<_DVORAK);
```
```suggestion
        set_single_persistent_default_layer(_DVORAK);
```

drashna commented:
Thanks!
